src/backend/access/table/table.c

```diff
@@ -209,8 +209,19 @@ CdbTryOpenTable(Oid relid, LOCKMODE reqmode, bool *lockUpgraded)
 	 * update or delete does not hold locks on catalog on segments, so
 	 * we do not need to consider lock-upgrade for DML on catalogs.
 	 */
+	/*
+	 * SINGLENODE_FIXME:
+	 * Since we don't need to dispatch in singlenode, we don't need to keep
+	 * the ExclusiveLock in singlenode mode.
+	 * However, it will cause behavior difference in `parallel_update_readcommitted`.
+	 * The second process in that case want to update the table is blocked but there is
+	 * `duplicate key value violates unique constraint "pg_aovisimap_95435_index"`
+	 * error after the first process commit.
+	 * Need to figure it out later. Currently, let's just use the stronger
+	 * lock as cbdb does.
+	 */
 	if (reqmode == RowExclusiveLock &&
-		Gp_role == GP_ROLE_DISPATCH &&
+		(Gp_role == GP_ROLE_DISPATCH || IS_SINGLENODE()) &&
```

Comment:
this line is buggy in currrent implementation or if we disabled utility login in "singlenode as a new role" scheme. `isolation` test suite use utility mode to login. in normal cases, `lockmode` are passed in as-is. but because of the failure in `uao/parallel_update_readcommitted`, locks are elevated from `RowExclusiveLock` to `ExclusiveLock` under singlenode mode, casing unnecessary blocking in `isolation` test suites. but if we remove this elevation behaviour, `uao/parallel_update_readcommitted` in `singlenode_isolation2` will failed. we could either disable tests in `isolation`, or `uao/parallel_update_readcommitted` in `singlenode_isolation2`, or trying to figure out why the latter failed. what do you think? @avamingli

Reply:
It sounds more like a bug, we'd better figure out.

Reply:
fixed in https://github.com/cloudberrydb/cloudberrydb/pull/206/commits/5e3d229d478a4ba4bddcab3d70baa8b8c614630a.